kratos/spatial_containers/bins_dynamic.h

```diff
@@ -36,232 +52,303 @@ std::size_t TDimension,
     class TDistanceIteratorType = typename std::vector<double>::iterator,
     class TDistanceFunction = Kratos::SearchUtils::SquaredDistanceFunction<TDimension,TPointType>
     >
-class BinsDynamic : public TreeNode<TDimension,TPointType, TPointerType, TIteratorType, TDistanceIteratorType, typename std::vector<TPointerType>::iterator >
+class BinsDynamic 
+    : public TreeNode<TDimension,TPointType, TPointerType, TIteratorType, TDistanceIteratorType, typename std::vector<TPointerType>::iterator >
 {
-
 public:
+    ///@name Type Definitions
+    ///@{
+
+    // Helper template to extract ObjectType or default to void
+    template <typename T, typename = void>
+    struct GetObjectType {
+        using type = void;
+    };
+
+    template <typename T>
+    struct GetObjectType<T, std::void_t<typename T::ObjectType>> {
+        using type = typename T::ObjectType;
+    };
 
     /// Pointer definition of BinsDynamic
     KRATOS_CLASS_POINTER_DEFINITION(BinsDynamic);
 
     enum { Dimension = TDimension };
 
-    typedef TPointType                                  PointType;
-    typedef TContainerType                              ContainerType;
-    typedef TIteratorType                               IteratorType;
-    typedef TDistanceIteratorType                       DistanceIteratorType;
-    typedef TPointerType                                PointerType;
-    typedef TDistanceFunction                           DistanceFunction;
-
-    typedef TreeNode<Dimension,TPointType,TPointerType,TIteratorType,TDistanceIteratorType> TreeNodeType;
+    using PointType = TPointType;
+    using ObjectType = typename GetObjectType<PointType>::type;
+    using ContainerType = TContainerType;
+    using IteratorType = TIteratorType;
+    using DistanceIteratorType = TDistanceIteratorType;
+    using PointerType = TPointerType;
+    using DistanceFunction = TDistanceFunction;
 
-    typedef typename TreeNodeType::CoordinateType       CoordinateType;  // double
-    typedef typename TreeNodeType::SizeType             SizeType;        // std::size_t
-    typedef typename TreeNodeType::IndexType            IndexType;       // std::size_t
+    using TreeNodeType = TreeNode<Dimension, TPointType, TPointerType, TIteratorType, TDistanceIteratorType>;
 
-    typedef Tvector<CoordinateType,Dimension>           CoordinateArray;
-    typedef Tvector<SizeType,Dimension>                 SizeArray;
-    typedef Tvector<IndexType,Dimension>                IndexArray;
+    using CoordinateType = typename TreeNodeType::CoordinateType; // double
+    using SizeType = typename TreeNodeType::SizeType;             // std::size_t
+    using IndexType = typename TreeNodeType::IndexType;           // std::size_t
```

Comment:
We should add a `using BoundingBoxType = BoundingBox<PointType>` so the user doesn't have to know the implementation while trying to call the `GetBoundingBox` he cannot use `auto`

Reply:
Yes